File: ios/SkyCastTests/Fakes.swift
import Foundation
@testable import SkyCast

// Hand-written test doubles.
//
// These are **fakes**, not mocks: each is a tiny working implementation whose state a test drives
// directly, and the compiler checks them against the protocol.
//
// Because view models depend only on `Domain` protocols, these fakes are all a view model test
// needs. No network, no SwiftData, no simulator services.

/// `WeatherRepository` whose emissions the test chooses.
final class FakeWeatherRepository: WeatherRepository, @unchecked Sendable {
    /// Emitted in order by `currentWeather(for:)`.
    var currentWeatherStates: [DataState<Weather>] = []
    var forecastStates: [DataState<Forecast>] = []

    /// Returned by `refresh(_:)`; set to a non-nil value to simulate a failed refresh.
    var refreshError: AppError?

    private(set) var refreshCallCount = 0
    private(set) var clearCacheCallCount = 0

    func currentWeather(for _: SavedLocation) -> AsyncStream<DataState<Weather>> {
        let states = currentWeatherStates
        return AsyncStream { continuation in
            for state in states {
                continuation.yield(state)
            }
            continuation.finish()
        }
    }

    func forecast(for _: SavedLocation) -> AsyncStream<DataState<Forecast>> {
        let states = forecastStates
        return AsyncStream { continuation in
            for state in states {
                continuation.yield(state)
            }
            continuation.finish()
        }
    }

    @discardableResult
    func refresh(_: SavedLocation) async -> AppError? {
        refreshCallCount += 1
        return refreshError
    }

    func clearCache() async {
        clearCacheCallCount += 1
    }
}

/// `LocationRepository` backed by an in-memory array.
final class FakeLocationRepository: LocationRepository, @unchecked Sendable {
    var locations: [SavedLocation] = []
    var searchResults: [LocationSearchResult] = []
    /// Thrown by `search(query:)` when set, so error paths are testable.
    var searchError: AppError?
    /// Thrown by `primaryLocation()` when set.
    var primaryLocationError: AppError?

    func savedLocations() async throws -> [SavedLocation] {
        locations.sorted { $0.sortOrder < $1.sortOrder }
    }

    func primaryLocation() async throws -> SavedLocation? {
        if let primaryLocationError {
            throw primaryLocationError
        }
        return locations.first(where: \.isPrimary)
    }

    func location(id: Int64) async throws -> SavedLocation? {
        locations.first { $0.id == id }
    }

    func search(query: String) async throws -> [LocationSearchResult] {
        if let searchError {
            throw searchError
        }
        guard query.trimmingCharacters(in: .whitespacesAndNewlines).count >= 2 else { return [] }
        return searchResults
    }

    @discardableResult
    func save(_ result: LocationSearchResult) async throws -> Int64 {
        let id = (locations.map(\.id).max() ?? 0) + 1
        locations.append(
            SavedLocation(
                id: id,
                name: result.name,
                countryCode: result.countryCode,
                state: result.state,
                latitude: result.latitude,
                longitude: result.longitude,
                sortOrder: locations.count,
                isPrimary: locations.isEmpty
            )
        )
        return id
    }

    func delete(_ location: SavedLocation) async throws {
        locations.removeAll { $0.id == location.id }
    }

    func setPrimary(_ location: SavedLocation) async throws {
        for index in locations.indices {
            locations[index].isPrimary = locations[index].id == location.id
        }
    }

    func reorder(ids: [Int64]) async throws {
        for (position, id) in ids.enumerated() {
            if let index = locations.firstIndex(where: { $0.id == id }) {
                locations[index].sortOrder = position
            }
        }
    }
}

/// `WeatherAPI` returning canned DTOs, for repository tests.
final class StubWeatherAPI: WeatherAPI, @unchecked Sendable {
    var currentWeatherResult: Result<CurrentWeatherDTO, Error> = .failure(AppError.offline)
    var forecastResult: Result<ForecastResponseDTO, Error> = .failure(AppError.offline)
    var searchResult: Result<[GeocodingResultDTO], Error> = .success([])

    private(set) var currentWeatherCallCount = 0

    func currentWeather(latitude _: Double, longitude _: Double) async throws -> CurrentWeatherDTO {
        currentWeatherCallCount += 1
        return try currentWeatherResult.get()
    }

    func forecast(latitude _: Double, longitude _: Double) async throws -> ForecastResponseDTO {
        try forecastResult.get()
    }

    func searchLocations(query _: String, limit _: Int) async throws -> [GeocodingResultDTO] {
        try searchResult.get()
    }
}

// MARK: - Sample data

enum Fixtures {
    /// A fixed instant, so no test depends on the wall clock.
    static let now = Date(timeIntervalSince1970: 1_785_931_200) // 2026-08-04T12:00:00Z

    static func location(
        id: Int64 = 1,
        name: String = "London",
        isPrimary: Bool = true
    )
        -> SavedLocation
    {
        SavedLocation(
            id: id,
            name: name,
            countryCode: "GB",
            state: "England",
            latitude: 51.5074,
            longitude: -0.1278,
            isPrimary: isPrimary
        )
    }

    static func weather(
        locationID: Int64 = 1,
        temperatureCelsius: Double = 22,
        cachedAt: Date = Fixtures.now
    )
        -> Weather
    {
        Weather(
            locationID: locationID,
            locationName: "London",
            condition: .clear,
            description: "Clear sky",
            iconCode: "01d",
            temperatureCelsius: temperatureCelsius,
            feelsLikeCelsius: temperatureCelsius - 1,
            minTemperatureCelsius: temperatureCelsius - 4,
            maxTemperatureCelsius: temperatureCelsius + 3,
            humidityPercent: 60,
            pressureHpa: 1_013,
            windSpeedMetresPerSecond: 4.5,
            windDirectionDegrees: 220,
            cloudinessPercent: 5,
            visibilityMetres: 10_000,
            sunrise: Fixtures.now.addingTimeInterval(-6 * 3_600),
            sunset: Fixtures.now.addingTimeInterval(8 * 3_600),
            observedAt: Fixtures.now,
            cachedAt: cachedAt
        )
    }

    static func searchResult(name: String = "London") -> LocationSearchResult {
        LocationSearchResult(
            name: name,
            countryCode: "GB",
            state: "England",
            latitude: 51.5074,
            longitude: -0.1278
        )
    }

    /// A minimal but realistic `/data/2.5/weather` payload, for decoding and mapper tests.
    static let currentWeatherJSON = """
    {
      "coord": { "lon": -0.1257, "lat": 51.5085 },
      "weather": [{ "id": 800, "main": "Clear", "description": "clear sky", "icon": "01d" }],
      "main": {
        "temp": 22.0, "feels_like": 21.4, "temp_min": 18.2,
        "temp_max": 24.6, "pressure": 1013, "humidity": 60
      },
      "visibility": 10000,
      "wind": { "speed": 4.5, "deg": 220 },
      "clouds": { "all": 5 },
      "dt": 1785931200,
      "sys": { "country": "GB", "sunrise": 1785909600, "sunset": 1785960000 },
      "timezone": 3600,
      "id": 2643743,
      "name": "London"
    }
    """
}

File: android/core/testing/src/main/kotlin/com/nauhaan/skycast/testing/Fakes.kt
package com.nauhaan.skycast.testing

import com.nauhaan.skycast.core.common.AppError
import com.nauhaan.skycast.core.common.DispatcherProvider
import com.nauhaan.skycast.core.common.NetworkMonitor
import com.nauhaan.skycast.domain.model.Forecast
import com.nauhaan.skycast.domain.model.LocationSearchResult
import com.nauhaan.skycast.domain.model.SavedLocation
import com.nauhaan.skycast.domain.model.TemperatureUnit
import com.nauhaan.skycast.domain.model.ThemeMode
import com.nauhaan.skycast.domain.model.UserPreferences
import com.nauhaan.skycast.domain.model.Weather
import com.nauhaan.skycast.domain.model.WeatherCondition
import com.nauhaan.skycast.domain.model.WindSpeedUnit
import com.nauhaan.skycast.domain.repository.DataState
import com.nauhaan.skycast.domain.repository.LocationRepository
import com.nauhaan.skycast.domain.repository.SettingsRepository
import com.nauhaan.skycast.domain.repository.WeatherRepository
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.map
import java.time.Instant
import java.time.ZoneOffset

/**
 * Hand-written test doubles.
 *
 * These are **fakes**, not mocks: each one is a tiny working implementation whose state
 * a test can drive directly. Fakes read better than `every { ... } returns ...` stacks
 * and they cannot drift out of sync with the interface, because the compiler checks them.
 *
 * They are also the payoff of the layering: because view models depend only on `domain`
 * interfaces, these fakes are all a view model test needs. No Robolectric, no device, no
 * network.
 */

class FakeWeatherRepository : WeatherRepository {
    val currentWeather = MutableStateFlow(DataState<Weather>())
    val forecast = MutableStateFlow(DataState<Forecast>())

    /** Returned by [refresh]; set to a non-null value to simulate a failed refresh. */
    var refreshError: AppError? = null

    var refreshCallCount = 0
        private set
    var clearCacheCallCount = 0
        private set

    override fun observeCurrentWeather(location: SavedLocation): Flow<DataState<Weather>> = currentWeather

    override fun observeForecast(location: SavedLocation): Flow<DataState<Forecast>> = forecast

    override suspend fun refresh(location: SavedLocation): AppError? {
        refreshCallCount++
        return refreshError
    }

    override suspend fun clearCache() {
        clearCacheCallCount++
    }
}

class FakeLocationRepository : LocationRepository {
    val savedLocations = MutableStateFlow<List<SavedLocation>>(emptyList())
    val primaryLocation = MutableStateFlow<SavedLocation?>(null)

    /** Thrown by [search] when set, so error paths are testable. */
    var searchError: AppError? = null
    var searchResults: List<LocationSearchResult> = emptyList()

    override fun observeSavedLocations(): Flow<List<SavedLocation>> = savedLocations

    override fun observePrimaryLocation(): Flow<SavedLocation?> = primaryLocation

    override suspend fun getById(id: Long): SavedLocation? = savedLocations.value.firstOrNull { it.id == id }

    override suspend fun search(query: String): List<LocationSearchResult> {
        searchError?.let { throw it }
        return searchResults
    }

    override suspend fun save(result: LocationSearchResult): Long {
        val id = (savedLocations.value.maxOfOrNull { it.id } ?: 0L) + 1
        val location =
            SavedLocation(
                id = id,
                name = result.name,
                countryCode = result.countryCode,
                state = result.state,
                latitude = result.latitude,
                longitude = result.longitude,
                sortOrder = savedLocations.value.size,
                isPrimary = savedLocations.value.isEmpty(),
            )
        savedLocations.value = savedLocations.value + location
        if (location.isPrimary) primaryLocation.value = location
        return id
    }

    override suspend fun delete(location: SavedLocation) {
        savedLocations.value = savedLocations.value - location
        if (primaryLocation.value == location) {
            primaryLocation.value = savedLocations.value.firstOrNull()
        }
    }

    override suspend fun setPrimary(location: SavedLocation) {
        primaryLocation.value = location
    }

    override suspend fun reorder(orderedIds: List<Long>) {
        savedLocations.value = orderedIds.mapNotNull { id ->
            savedLocations.value.firstOrNull { it.id == id }
        }
    }
}

class FakeSettingsRepository : SettingsRepository {
    val preferences = MutableStateFlow(UserPreferences())

    override fun observePreferences(): Flow<UserPreferences> = preferences

    override suspend fun setTemperatureUnit(unit: TemperatureUnit) {
        preferences.value = preferences.value.copy(temperatureUnit = unit)
    }

    override suspend fun setWindSpeedUnit(unit: WindSpeedUnit) {
        preferences.value = preferences.value.copy(windSpeedUnit = unit)
    }

    override suspend fun setThemeMode(mode: ThemeMode) {
        preferences.value = preferences.value.copy(themeMode = mode)
    }

    override suspend fun setUseDynamicColour(enabled: Boolean) {
        preferences.value = preferences.value.copy(useDynamicColour = enabled)
    }

    override suspend fun reset() {
        preferences.value = UserPreferences()
    }
}

class FakeNetworkMonitor(initiallyOnline: Boolean = true) : NetworkMonitor {
    val online = MutableStateFlow(initiallyOnline)
    override val isOnline: Flow<Boolean> = online.map { it }
}

/** Routes every dispatcher to the test dispatcher so coroutines are deterministic. */
class TestDispatcherProvider(private val dispatcher: CoroutineDispatcher) : DispatcherProvider {
    override val main: CoroutineDispatcher = dispatcher
    override val io: CoroutineDispatcher = dispatcher
    override val default: CoroutineDispatcher = dispatcher
}

// ── Sample data ────────────────────────────────────────────────────────────

/** A fixed instant, so no test depends on the wall clock. */
val TestInstant: Instant = Instant.parse("2026-08-04T12:00:00Z")

fun sampleLocation(id: Long = 1L, name: String = "London", isPrimary: Boolean = true): SavedLocation = SavedLocation(
    id = id,
    name = name,
    countryCode = "GB",
    state = "England",
    latitude = 51.5074,
    longitude = -0.1278,
    isPrimary = isPrimary,
)

fun sampleWeather(
    locationId: Long = 1L,
    temperatureCelsius: Double = 22.0,
    cachedAt: Instant = TestInstant,
    // Overridable so a test can assert that sunrise/sunset render in the *location's* zone
    // rather than the JVM's, which is the one thing a UTC-only fixture cannot distinguish.
    zoneOffset: ZoneOffset = ZoneOffset.UTC,
): Weather =
    Weather(
        locationId = locationId,
        locationName = "London",
        condition = WeatherCondition.CLEAR,
        description = "Clear sky",
        iconCode = "01d",
        temperatureCelsius = temperatureCelsius,
        feelsLikeCelsius = temperatureCelsius - 1,
        minTemperatureCelsius = temperatureCelsius - 4,
        maxTemperatureCelsius = temperatureCelsius + 3,
        humidityPercent = 60,
        pressureHpa = 1013,
        windSpeedMetresPerSecond = 4.5,
        windDirectionDegrees = 220,
        cloudinessPercent = 5,
        visibilityMetres = 10_000,
        sunrise = TestInstant.minusSeconds(6 * 60 * 60),
        sunset = TestInstant.plusSeconds(8 * 60 * 60),
        observedAt = TestInstant,
        cachedAt = cachedAt,
        zoneOffset = zoneOffset,
    )
